tests/components/tracked-controls.test.js

```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
+    });
+
+    test('emits independent streams for buttonup, touchend and buttonchanged', function () {
+      const emitSpy = this.sinon.spy(el, 'emit');
+      component.buttonStates[0] = {pressed: true, touched: true, value: 1};
+      controller.buttons[0].pressed = false;
+      controller.buttons[0].touched = false;
+      controller.buttons[0].value = 0;
+      component.tick();
+
+      // Filter down to just our expected events for verification.
+      const emitButtonCalls = emitSpy.getCalls().filter(call => call.args[0] === 'buttonup');
+      assert.equal(emitButtonCalls.length, 1);
+
+      const emitTouchCalls = emitSpy.getCalls().filter(call => call.args[0] === 'touchend');
+      assert.equal(emitTouchCalls.length, 1);
+
+      const emitChangedCalls = emitSpy.getCalls().filter(call => call.args[0] === 'buttonchanged');
+      assert.equal(emitChangedCalls.length, 1);
+
+      // Verify each of the 3 events has the correct event name and state.
+      assertButtonEvent(emitButtonCalls[0], 'buttonup', 0, controller.buttons[0]);
+      assertButtonEvent(emitTouchCalls[0], 'touchend', 0, controller.buttons[0]);
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
+    });
+
+    test('emits correct event stream for spaced out interaction.', function () {
+      // First round, verify we only see a touchstart and buttonchanged
+      let emitSpy = this.sinon.spy(el, 'emit');
+      controller.buttons[0].touched = true;
+      component.tick();
+
+      assertEventStream(emitSpy.getCalls(), ['buttonchanged', 'touchstart'], ['buttondown', 'buttonup', 'touchend']);
+      emitSpy.restore();
+
+      // Second round, verify we only see a buttondown and buttonchanged since pressed state isn't changing.
+      emitSpy = this.sinon.spy(el, 'emit');
+      controller.buttons[0].pressed = true;
+      component.tick();
+
+      assertEventStream(emitSpy.getCalls(), ['buttonchanged', 'buttondown'], ['buttonup', 'touchend', 'touchstart']);
+      emitSpy.restore();
+
+      // Third round, verify we only see a buttonup and button changed when we release the button.
+      emitSpy = this.sinon.spy(el, 'emit');
+      controller.buttons[0].pressed = false;
+      component.tick();
+
+      assertEventStream(emitSpy.getCalls(), ['buttonchanged', 'buttonup'], ['buttondown', 'touchend', 'touchstart']);
+      emitSpy.restore();
+
+      // Fourth round, verify we only see a touchend and button changed when we lift our touch.
+      emitSpy = this.sinon.spy(el, 'emit');
+      controller.buttons[0].touched = false;
+      component.tick();
+
+      assertEventStream(emitSpy.getCalls(), ['buttonchanged', 'touchend'], ['buttondown', 'buttonup', 'touchstart']);
+      emitSpy.restore();
+    });
+
+    test('emits correct event states on a fast click', function () {
```

Comment:
Verifies that if the states all change in single ticks that we still get proper event streams. Special case of the prior test that also tests that we have independent event streams.